### PR TITLE
[InstCombine] Do not simplify lshr/shl arg if it is part of fshl rotate pattern.

### PR DESCRIPTION
**Pull request by quic-eikansh**
The fshl/fshr having first two arguments as same gets lowered to targets
specific rotate. But based on the uses, one of the arguments can get
simplified resulting in different arguments performing equivalent operation.

This patch prevents the simplification of the arguments of lshr/shl if they are
part of fshl pattern.

### CODE REVIEW

**Comment by quic-eikansh** [comment]
This pull request is created to address the reviews in https://github.com/llvm/llvm-project/pull/66115

**Comment by goldsteinn** [comment]
I would expect to see test difference in the commit adding the new restriction.

**Comment by quic-eikansh** [comment]
> I would expect to see test difference in the commit adding the new restriction.

I didn't get what you meant by this. Should I have code changes and test in the same commit?

**Comment by goldsteinn** [comment]
> > I would expect to see test difference in the commit adding the new restriction.
> 
> I didn't get what you meant by this. Should I have code changes and test in the same commit?

Yes, otherwise the commit with the impl will not be passing (and it makes review more difficult as its hard to see the proper side affects of a change).

**Comment by quic-eikansh** [comment]
I have squashed the changes and test into one commit. @goldsteinn

**Comment by goldsteinn** [comment]
> I have squashed the changes and test into one commit. @goldsteinn

Thats not really ideal either... Then its difficult to review the test changes.
Typically the way we do it is two seperate commits i.e

commit 1: Tests (w/ checks before the implementation)
commit 2: Implementation (re update test checks)

That way in commit 2 we can see what actual changes the implementation is causing.

**Comment by quic-eikansh** [comment]
@goldsteinn I have made the changes as you suggested.

**Comment by quic-eikansh** [comment]
@goldsteinn Did you get a chance to look at it?

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:2711]:
```suggestion
  assert(Or.getOpcode() == BinaryOperator::Or && "Expecting or instruction");
  // TODO: Can we reduce the code duplication between this and the related
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (quic-eikansh) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:2709]: Do `std::pair`  has advantage over `std::tuple`? I see `std::tuple` used in codebase even for 2 element. I have addressed other 2 reviews.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thanks!
Please wait for additional approval from other reviewers.

**Comment by quic-eikansh** [comment]
@goldsteinn @nikic

**Comment by quic-eikansh** [comment]
> LGTM. Thanks! Please wait for additional approval from other reviewers.

Thanks

**Comment by dtcxzyw** [comment]
Ping.

**Review by efriedma-quic — APPROVED** [review]
Changes LGTM... but I'd like to land as a series of three commits; one for the refactor (6bbc776 and the fixup), one for the new tests (b896ff5), then one for the actual behavior change.

Please push separate PRs for the refactor and the test, then I'll merge everything for you.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:2709]: I agree that we should use std::pair instead of std::tuple if there are only two elements.

**Pull request closed without merging**
